### ISSUE TITLE
Infinite loop in try_push_visible_item_path 

### ISSUE THREAD

**Issue by hvenev** [rust-lang/rust#57500]
This happens when an item has an infinite number of visible names:

```
// pathlooptest/src/lib.rs
pub struct AStruct;

pub mod prelude {
    pub use crate as pathlooptest;
    pub use crate::AStruct;
}
```

```
// main.rs
use pathlooptest::prelude::*;

pub fn main() {
    let x: AStruct = 42; // causes error, all public names of AStruct are generated?
}
```

**Comment by hvenev** [comment]
The hang can also occur on code that should produce no errors/warnings. However I couldn't find out why.

**Comment by jakubadamw** [comment]
This seems to be fixed. I verified both with the test case above and using the repro script from https://github.com/rust-lang/rust/issues/58430#issue-409902389.

@rustbot modify labels to +E-needs-test
